src/vm.c

```diff
@@ -72,6 +76,87 @@ static void vm_init_cpu_id(vm_t *v)
     ioctl(v->vcpu_fd, KVM_SET_CPUID2, &kvm_cpuid);
 }
 
+#define FREQ_NS 1000000
+
+/* global state to stop the loop of thread */
+static volatile bool thread_stop = false;
+static pthread_t main_thread;
+
+static void *timer_thread(void *args)
+{
+    struct epoll_event events[1];
+    int epoll_fd = *(int *) args;
+
+    while (!__atomic_load_n(&thread_stop, __ATOMIC_RELAXED)) {
+        int nfds = epoll_wait(epoll_fd, events, 1, -1);
+        if (nfds < 0)
+            continue;
+
+        assert(nfds == 1);
+
+        uint64_t exp;
+        ssize_t size = read(events[0].data.fd, &exp, sizeof(uint64_t));
+        if (size != sizeof(uint64_t))
+            return NULL;
+
+        pthread_kill((pthread_t) main_thread, SIGUSR1);
+    }
+
+    return NULL;
+}
+
+static void handler(int sig, siginfo_t *si, void *uc) {}
+
+static int vm_init_timer(vm_t *v)
+{
+    int epoll_fd, timer_fd;
+    struct epoll_event ev;
+    struct itimerspec its;
+    struct sigaction sa;
+    sigset_t mask;
+
+    sa.sa_flags = SA_SIGINFO;
+    sa.sa_sigaction = handler;
+    sigemptyset(&sa.sa_mask);
+    if (sigaction(SIGUSR1, &sa, NULL) == -1)
+        return throw_err("Failed to create signal handler");
+
+    /* Block timer signal temporarily. */
+    sigemptyset(&mask);
+    sigaddset(&mask, SIGUSR1);
+    if (sigprocmask(SIG_SETMASK, &mask, NULL) == -1)
+        return throw_err("Failed to block timer signal");
+
+    epoll_fd = epoll_create1(0);
+    if (epoll_fd == -1)
+        return throw_err("Failed to created epoll fd");
+
+    timer_fd = timerfd_create(CLOCK_MONOTONIC, TFD_CLOEXEC | TFD_NONBLOCK);
```

Comment:
Change to `CLOCK_MONOTONIC_RAW` since we don't care about NTP.

Reply:
`CLOCK_MONOTONIC_RAW` is the invalid value for timerfd, see [fs/timerfd.c](https://elixir.bootlin.com/linux/latest/source/fs/timerfd.c#406)